climate_index_collection/reductions.py

```diff
@@ -428,3 +433,79 @@ def eof_weights(dobj):
         Square root of weights needed to pre-process input data for SVD.
     """
     return np.sqrt(np.cos(np.deg2rad(dobj.coords["lat"])))
+
+
+def polygon_prime_meridian(pg):
+    """
+    Transforms shapely Polygons or MultiPolygons defined in [180W, 180E) coords into [0E,360E) coords.
+    Takes care of Polygons crossing the prime meridan.
+    Polygon points are expected be (lon, lat) tuples.
+
+    Parameters
+    ----------
+    pg: shaply Polygon or shapely MultiPolygon
+        Polygon including the area wanted.
+
+    Returns
+    -------
+    shapely MultPolygon
+        shaply MultiPolygon with single.
+    """
+
+    # create prime_meridian to eventually split the polygon
+    prime_meridian = LineString([(0, 90), (0, -90)])
+    pg_split = split(pg, prime_meridian)
+    # create a list containing all polygons
+    pg_list = []
+    for temp_pg in pg_split.geoms:
+        # check if the poygon minx is negative and and 360deg to it.
+        if temp_pg.bounds[0] < 0:
+            temp_pg = translate(temp_pg, xoff=360)
+        pg_list += [temp_pg]
+    # create the multipolygon existing in [0E, 360E) coords from the list of polygons
+    return unary_union(pg_list)
+
+
+def polygon2mask(dobj, pg, lat_name="lat", lon_name="lon"):
+    """
+    This funciton creates a mask for a given DataArray or DataSet based on a shapely Polygon or MultiPolygon.
+    Polygon points are expected be (lon, lat) tuples.
+    The dobj is expected to have lon values in [0E, 360E) coords.
+
+    Parameters
+    ----------
+    dobj: xarray.Dataset or xarray.DataArray
+        Contains the original data.
+    pg: shaply Polygon or shapely MultiPolygon
+        Polygon including the area wanted.
+    lat_name: str
+        Name of the latitude coordinate. Defaults to "lat".
+    lon_name: str
+        Name of the longitude coordinate. Defaults to "lon".
+
+    Returns
+    -------
+    xarray.Dataset or xarray.DataArray
+        Mask for given Polygon on dobj grid.
+    """
+
+    # handle prime meridian crossing Polygons and transform [180W, 180E) coords into [0E,360E) coords
```

Comment:
We could think about something that also works for other representations of the prime meridian (any of `[(lon_prime, -90), (lon_prime, 90)]` with `lon_prime =  ..., -360, 0, +360, ...`. This should be easy to do, because we just need to find the min and max lon in the Polygon and then cut with every prime meridian we need. And we'd need to adapt the shifting into the `[0, 360)` base interval.

Reply:
Yeah that is a good idea!
So we would need in fact two "prime meridans" one a each side of the lon grid, am I right on that?
Do you think this can handle non rectangular grids?

Reply:
But not sure if this is on my top priority at the moment.
Might be good to at least make a partial merge in advance, so we can get the AMO index working.

Reply:
Yes, let's keep this in mind for later.

> need in fact two "prime meridans"

No, we'd need as many as we'd need. A region thats spiralling around the globe from 730W to 1090E could also be handled by just splitting at -720, -360, 0, 360, 720, 1080.

Not sure what you mean by non-rectangular grids.

Reply:
I've opened #116 to not forget about this idea. Marking this conversation as resolved.